fetch/mesh_data_test.go

```diff
@@ -1038,6 +1040,38 @@ func Test_GetAtxsLimiting(t *testing.T) {
 			require.NoError(t, f.Start())
 			t.Cleanup(f.Stop)
 
+			// Connect the P2P mesh only after the server is configured.
+			// This way, we avoid the race causing bad protocol identification.
+			require.NoError(t, mesh.ConnectAllButSelf())
+			// There are possibilities for race in go-libp2p when we connect early
+			// after server creation. IDService has its own snapshot which is updated
+			// upon EvtLocalProtocolsUpdated event. The events propagate over a channel
+			// and there's no way to see if IDService has already updated its snapshot.
+			// In case if the snapshot is outdated, a wrong list of supported protocols
+			// can be sent to the peers. The only way around it is apparently to re-connect
+			// if the necessary protocol is not supported.
+			require.Eventually(t, func() bool {
+				if ps.Total() == 0 {
+					return false
+				}
+				if len(ps.SelectBestWithProtocols(1, []protocol.ID{
+					hashProtocol,
+				})) != 0 {
+					return true
+				}
+				nets := mesh.Nets()
+				for _, n1 := range nets {
+					for _, n2 := range nets {
+						if n1 == n2 {
+							continue
+						}
+						require.NoError(t, mesh.DisconnectNets(n1, n2))
+					}
+				}
+				require.NoError(t, mesh.ConnectAllButSelf())
+				return false
+			}, 1*time.Second, 5*time.Millisecond)
```

Comment:
Could this be a problem in other tests or even the actual code as well?

Reply:
In theory, yes; in practice, I tried to verify other places where such fake mesh network is used with `stress` and so far failed to stumble upon any failures. We probably should use this approach if we have any problems with other such tests, but I'm unsure we should do it right away

Reply:
Re `IdentifyWait` vs `IdentifyConn`: the channel that `IdentifyWait` returns will be closed at most 30s after `IdentifyWait` is called. I replaced `IdentifyConn` with `IdentifyWait` and added 3s timeout, as if a peer is that slow it's not feasible to fetch from anyway. I could also add a context to the function that retrieves the list of protocols, but given the current fetcher codebase it will probably only add some unneeded complication